crates/autopilot/src/infra/solvers/authenticator.rs

```diff
@@ -0,0 +1,107 @@
+use {
+    crate::{
+        domain::{self, eth},
+        infra::blockchain::{contracts::deployment_address, ChainId},
+    },
+    ethcontract::dyns::DynWeb3,
+    primitive_types::H160,
+};
+
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+pub struct Authenticator {
+    /// The authenticator contract used for allow-listing solvers to settle.
+    authenticator: contracts::GPv2AllowListAuthentication,
+    /// The safe module that is used to provide special role to EOA.
+    authenticator_role: contracts::Roles,
+    /// The EOA that is allowed to add/remove solvers.
+    authenticator_eoa: ethcontract::Account,
+}
+
+pub struct Addresses {
+    pub settlement: Option<eth::Address>,
+    pub authenticator_eoa: eth::H256,
+}
+
+impl Authenticator {
+    /// Creates an authenticator which can remove solvers from the allow-list
+    pub async fn new(web3: DynWeb3, chain: ChainId, addresses: Addresses) -> Self {
+        let address_for = |contract: &ethcontract::Contract, address: Option<H160>| {
+            address
+                .or_else(|| deployment_address(contract, &chain))
+                .unwrap()
+        };
+
+        let settlement = contracts::GPv2Settlement::at(
+            &web3,
+            address_for(
+                contracts::GPv2Settlement::raw_contract(),
+                addresses.settlement.map(Into::into),
+            ),
+        );
+
+        let authenticator = contracts::GPv2AllowListAuthentication::at(
+            &web3,
+            settlement
+                .authenticator()
+                .call()
+                .await
+                .expect("authenticator address"),
+        );
+
+        let authenticator_role = contracts::Roles::at(
```

Comment:
Also why do we effectively overriding the authenticator address but not the roles address? IMO another indicator that these dependencies should be injected.

Reply:
True. IMO all contracts that the `Authenticator` depends on need to be received via `Addresses`.

Reply:
but in that case shouldn't we be passing `Contracts` instead of `Addresses`? does it also make sense for the addresses which are specific for the circuit breaker? like the role or authenticator eoa?

Reply:
If we want to make the component as independent as possible, it should receive raw configuration params so it can independently build whatever it needs to work properly.

Reply:
With that in mind, `Contracts` is an internal autopilot component and should be forwarded to `Authenticator`. The reason being once we wish to move `Authenticator` out of the `autopilot` there would be no need to refactor constructor to not work with `Contracts`.

Reply:
but I think that making an object depending on another object within the same crate is totally ok. In my mind, having a struct which instantiates the contracts and then those contracts are use somewhere else (in circuit breaker) would make sense :thinking: 

If you want to make the module as independent as possible, then we can have two builders (or `new`) for it, once with raw address and the other with the contract, so if we move it out of autopilot it won't require any change.

Reply:
I don't want to bikeshed on this too much, since it's not essential for delivery of feature (and it also won't be too problematic to refactor this if needed). That said, I am ok with both versions.